docs: Extend BPF-based masquerading section

brb opened the pull request:
This PR:

- Extend the section.
- Extends `cilium status` to show masq devices.
- Improve usage of `global.devices` in the kubeproxy-free gsg.

Reviewable per commit.

brb committed:
docs: Extend BPF masquerading section

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
cli: Report BPF masquerading devices

This commit extends "cilium status" to show which devices can run
the BPF masquerading program. E.g.:

    $ cilium status | grep Masquerading
    Masquerading:           BPF   [eth0, eth1]

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
docs: Improve usage of global.devices in kubeproxy-free gsg

Signed-off-by: Martynas Pumputis <m@lambda.lt>

joestringer commented on Documentation/concepts/networking/masquerading.rst:
```diff
@@ -52,9 +52,90 @@ Implementation Modes
+
+The eBPF-based implementation is the most efficient
+implementation. It requires Linux kernel 4.19 and can be enabled with
+the ``global.bpfMasquerade=true`` helm option (enabled by default).
```
Note that the previous section describes from a high-level how to set the masquerade interface, and it implies that iptables is required to configure masquerading interfaces from the sentence "This setting is only available in....". I think we should (even if briefly) expand the above section to point out that there is a BPF-based implementation as well which is configured the devices flag.

brb replied:
That's a good spot. As configuring the masquerading interface depends on the implementation, I moved the `--egress-masquerade-interface` bits to the `iptables-based` section instead.

joestringer commented on Documentation/concepts/networking/masquerading.rst:
```diff
@@ -52,9 +52,90 @@ Implementation Modes
+specified, the program will be automatically attached to the devices selected by
+:ref:`the BPF NodePort device detection metchanism <Nodeport Devices>`.
+To manually change this, use the ``global.devices`` helm option. ``cilium status``
+can be used to determine on which devices the program is running:
```
nit: 
```suggestion
To manually change this, use the ``global.devices`` helm option. Use ``cilium status``
determine which devices the program is running on:
```

joestringer commented on Documentation/concepts/networking/masquerading.rst:
```diff
@@ -52,9 +52,90 @@ Implementation Modes
+To manually change this, use the ``global.devices`` helm option. ``cilium status``
+can be used to determine on which devices the program is running:
+
+.. parsed-literal::
```
This command doesn`t parse any literals that I can see, it's sufficient to either:

```suggestion
.. code:: shell-session
```

or 

```suggestion
::
```

joestringer replied:
(Note that `shell-session` highlighting is only useful if you're actually starting command lines with `$` / `#`; I'm not sure we've consistently applied a ruling on how to define these across all of the examples by this point..)

joestringer commented on Documentation/concepts/networking/masquerading.rst:
```diff
@@ -52,9 +52,90 @@ Implementation Modes
+- ICMP (only Echo request and Echo reply)
+
+By default, any packet from a pod destined to outside of the ``native-routing-cidr``
+range is masqueraded. To allow more fine-grained control, Cilium implements
```
```suggestion
By default, any packet from a pod destined to an IP address outside of the ``native-routing-cidr``
range is masqueraded. To allow more fine-grained control, Cilium implements
```

joestringer commented on Documentation/concepts/networking/masquerading.rst:
```diff
@@ -52,9 +52,90 @@ Implementation Modes
+In addition, if the ``masqLinkLocal`` is not set or set to false, then
+``169.254.0.0/16`` is appended to the non-masquerade CIDRs list.
+
+The agent uses Fsnotify to track updates to the configuration file, so the original
+``resyncInterval`` option is unnecessary.
+
```
```suggestion
```

If it's unnecessary, then why do we document it?

brb replied:
Because the vanilla ip-masq-agent requires this option, and I wanted to make clear that we don't poll the config file.

joestringer commented on Documentation/spelling_wordlist.txt:
```diff
@@ -245,6 +246,7 @@ committer
 committers
 conf
 config
+configmap
```
nit: I suspect that if you use `ConfigMap` in single back-ticks rather than plain then you won't need this.

brb replied:
Aye, you are right;-)

joestringer reviewed (changes requested):
Awesome thanks! A few minor nits below.

brb committed:
datapath: Add GetRemoteSNATDstAddrExclusionCIDR helper

The helper is used to determine the dst CIDR for SNAT exclusion.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
api,daemon,cli: Add SNAT exclusion dst to cilium status

This commits extends "cilium status" to show dst cidr of SNAT exclusion.
E.g.:

    $ cilium status | grep Masquerading
    Masquerading:           BPF   [eth0, eth1]  10.0.0.0/16

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
docs: Add exclusion CIDR output to BPF masq section

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm reviewed (approved):
I didn't read the docs only code.

joestringer commented on Documentation/concepts/networking/masquerading.rst:
```diff
@@ -32,29 +32,111 @@ Setting the routable CIDR
+``resyncInterval`` option is unnecessary.
+
+I think it's important for users to understand how we sync a config. The example below
+shows how to configure the agent via `ConfigMap` and to verify it:
```
```suggestion
The example below shows how to configure the agent via `ConfigMap` and to verify it:
```

brb replied:
Oh boy (accidental paste)...

joestringer commented on Documentation/concepts/networking/masquerading.rst:
```diff
@@ -32,29 +32,111 @@ Setting the routable CIDR
+I think it's important for users to understand how we sync a config. The example below
+shows how to configure the agent via `ConfigMap` and to verify it:
+
+::
```
Any reason to avoid using `.. code-block:: shell-session` and obvious shell bits like starting lines you execute with `$`? Seems a bit weird to not differentiate the commands and the output here.

brb replied:
Agree, but I wanted to make it consistent with other guides. IIRC we had a discussion some time ago whether we should keep `$`'s in our guides. The consensus was to get rid of them.

joestringer commented on pkg/datapath/config.go:
```diff
@@ -120,3 +122,15 @@ type ConfigWriter interface {
+// GetRemoteSNATDstAddrExclusionCIDR returns a CIDR for SNAT exclusion. Any
+// packet sent from a local endpoint to an IP address belonging to the CIDR
+// should not be SNAT'd.
+func GetRemoteSNATDstAddrExclusionCIDR() *cidr.CIDR {
```
nit: No need to `Get..()`; AFAIK usually we avoid this in Go unless there's ambiguity with a public field of the same name.

```suggestion
func RemoteSNATDstAddrExclusionCIDR() *cidr.CIDR {
```

joestringer commented on pkg/datapath/config.go:
```diff
@@ -18,8 +18,10 @@ import (
+	"github.com/cilium/cilium/pkg/cidr"
 	"github.com/cilium/cilium/pkg/identity"
 	"github.com/cilium/cilium/pkg/mac"
+	"github.com/cilium/cilium/pkg/node"
```
I have this sneaking suspicion we're trying to avoid `node` dependency in the `pkg/datapath` so it's mainly a types package.... but I guess if it's a problem, we can always figure out how to rearrange it.

joestringer replied:
While we could remove this import by pushing the implementation to `pkg/node`, I don't think it makes much logical sense there given that this is about how we choose to SNAT or not. I have no better immediate suggestion than keeping the current version.

joestringer commented on pkg/datapath/linux/config/config.go:
```diff
@@ -334,12 +333,7 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
-			} else {
-				cidr = node.GetIPv4AllocRange()
-			}
+			cidr := datapath.GetRemoteSNATDstAddrExclusionCIDR()
```
nice spotting :)

joestringer reviewed (changes requested):
Minor nits, otherwise LGTM.
